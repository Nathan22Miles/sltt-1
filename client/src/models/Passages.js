import { types, getParent, detach } from "mobx-state-tree"
import _ from 'underscore'
import { _insertBy, _remove } from './ModelUtils.js'

import { checkStatus, pushFile, concatBlobs, getUrl } from './API.js'
import { user } from '../components/auth/User.js'
import { deletedStatus } from './PassagesStatus.js'

const debug = require('debug')('sltt:Passages') 

let previousDate = new Date()

function formatTimestamp(date) {
    let ts = date.toISOString().slice(0, 19)
    ts = ts.replace(/:/g, '.')
    ts = ts.replace('T', ' ')
    return ts
}

// Generate a timestamp. Make sure it is not same as previously generated timestamp.
export const timestamp = function() {
    let nextDate = new Date()
    if (nextDate - previousDate < 1000) {
        nextDate = new Date(previousDate.getTime() + 1000)
    }
    previousDate = nextDate
    return formatTimestamp(nextDate)
}

function uploadTimestamp(file) {
    return formatTimestamp(new Date(file.lastModified))
}


const setSignedUrl = function(setter, project, url, cb) {
    getUrl(project, url)
        .then(signedUrl => {
            debug(`setSignedUrl ${signedUrl.slice(0,40)}`)
            setter(signedUrl)
            cb && cb()
        })
        .catch(
            err => { cb && cb(err) }
        )
}

// A PassageNoteSegment represents one comment in a note thread.
// The comment can be a video with a 'duration' and 'url'.
// The comment can be textual with 'text' (draft-js content)

export const PassageNoteSegment = types.model("PassageNoteSegment", {
    _id: types.string,
    videoCreated: types.string,
    noteCreated: types.string,
    segmentCreated: types.string,
    username: types.string,
    position: types.number,
    duration: types.number,
    url: types.string,
    signedUrl: types.optional(types.string, ''),    
    text: types.string,

    // videoCreated, noteCreated, segmentCreated, username, position, duration, text, url, 
})
.actions(self => ({
    _getSignedUrl: () => {
        return new Promise((resolve, reject) => {
            let projectName = self.getPassage().getPortion().getProject().name
            setSignedUrl(self.setSignedUrl, projectName, self.url, err => {
                if (err) reject(err)
                else resolve()
            })
        })
    },

    getPassage: () => {
        let passage = getParent(self, 4)
        return passage
    },

    setSignedUrl: signedUrl => {
        self.signedUrl = signedUrl
    },
}))


export const PassageNote = types.model("PassageNote", {
    noteCreated: types.identifier(),
    videoCreated: types.string,
    position: types.number,
    segments: types.array(PassageNoteSegment), 
    resolved: types.boolean,
})
.actions(self => ({
    getSignedUrls: (cb) => {
        let _gets = _.map(self.segments, segment => segment._getSignedUrl())
        Promise.all(_gets)
            .then(() => {cb && cb()})
            .catch(err => {cb && cb(err)})
    },

    apply: (doc) => { 
        if (doc.tag === 'resolved') {
            self.resolved = true
            return
        }

        let { segmentCreated } = doc
        let segment = _.findWhere(self.segments, { segmentCreated })
        if (segment) return   // Segment already exists in note, nothing to do

        _insertBy(doc, self.segments, 'segmentCreated')
    },

    resolve: (cb) => {
        if (self.resolved) {
            cb && cb()
            return
        }

        self.resolved = true

        let doc = {
            videoCreated: self.videoCreated,
            noteCreated: self.noteCreated,
            position: self.position,
            segmentCreated: timestamp(),
        }

        let passage = getParent(self, 2)
        passage.setId(doc, 'resolved')
        passage.put(doc, cb)        
    },

    addSegment: (doc, cb) => {
        let passage = getParent(self, 2)
        passage.setId(doc, 'notesegment', 'webm')
        self.insertSegment(doc)
        passage.put(doc, cb)
    },

    getDb: () => {
        let passage = getParent(self, 2)
        let _db = passage.getPortion().getDb()
        return _db
    },

    insertSegment(doc) {
        let segment =_insertBy(doc, self.segments, 'segmentCreated')
        segment._getSignedUrl().catch(err => { debug('!!_getSignedUrl failed') })
    },

    removeSegment: (_id, cb) => {
        let _db = self.getDb()
        _remove(_db, self.segments, _id, cb)
    },
}))

export const PassageStatus = types.model("PassageStatus", {
    statusCreated: types.string,
    videoCreated: types.string,
    status: types.number,
})
.actions(self => ({
    setStatusCreated: statusCreated => {
        self.statusCreated = statusCreated
    },
}))


export const PassageVideo = types.model("PassageVideo", {
    _id: types.identifier(),
    username: types.string,
    videoCreated: types.string,
    duration: types.number,
    url: types.string,
    signedUrl: types.optional(types.string, ''),
})
.actions(self => ({

    // Get the signedUrl for this video into the signedUrl attribute.
    // If cb is passed, do this as a callback.
    // Otherwise return a promise.
    getSignedUrl: (cb) => {
        let projectName = self.getPassage().getPortion().getProject().name

        if (cb) {
            setSignedUrl(self.setSignedUrl, projectName, self.url, cb)
            return null
        }

        return new Promise((resolve, reject) => {
            setSignedUrl(self.setSignedUrl, projectName, self.url, err => {
                if (err) reject(err)
                else resolve()
            })
        })
    },

    getPassage: () => {
        return getParent(self, 2)
    },

    setStatus: status => {
        self.getPassage().setStatus(self, status)
    },

    setSignedUrl: signedUrl => {
        //debug('PassageVideo setSignedUrl', signedUrl && signedUrl.substring(0, 40))
        self.signedUrl = signedUrl
    },
}))
.views(self => ({
    get created() {
        return self.videoCreated.slice(0, -3)
    },

    get status() {
        // Statuses is an array of status setting transactions.
        // Setting a status appends a new transaction to the array.
        // We do this because offline operation is made more stable by only adding to the
        // data model, never changing an existing item.
        // Because of this we have to find the latest transaction correspond to this videoPassage.

        let statuses = self.getPassage().statuses
        statuses = _.where(statuses, {videoCreated: self.videoCreated})
        if (statuses.length === 0) return 0

        statuses = _.sortBy(statuses, 'statusCreated')
        return _.last(statuses).status
    },

    get isLatest() {
        let passage = self.getPassage()
        let last = _.last(passage.videos)
        return self._id === last._id
    },
}))

export const Passage = types.model("Passage", {
    _id: types.identifier(),
    name: types.string,
    rank: types.number,
    statuses: types.array(PassageStatus),
    notes: types.array(PassageNote),
    videos: types.array(PassageVideo),
})
.actions(self => ({
    videod: () => self.videos.length > 0,

    getPortion: () => getParent(self, 2),

    //!! Need a unit test case for this.
    uploadFile: (file, onprogress, cb) => {
        let portion = self.getPortion()
        let projectName = portion.getProject().name
        let videoCreated = uploadTimestamp(file)

        let path = `${portion.name}/${self.name}/${videoCreated}`

        pushFile(file, projectName, path, onprogress)
            .then(checkStatus)
            // WARNING, if I replace the following with just '.then(concatBlobs(...)')
            // it does not invoke concatBlobs promise.
            // Perhaps that means you have to return a Promise to get it invoked.
            // You cannot just pass a Promise to then(...) ???
            .then(() => concatBlobs(projectName, path, 1))
            .then(checkStatus)
            .then(response => response.text())
            .then(url => {
                let doc = {
                    username: user.username,
                    videoCreated,
                    duration: 0, // We don't know the duration yet of an uploaded video
                    url,
                }

                self.setId(doc, 'video', 'mp4')
                self.pushVideo(doc)

                return self.putPromise(doc)
            })
            .then(() => {
                cb()
            })
            .catch(err => { 
                cb(err) 
            })
    },

    pushVideo: (doc) => {
        self.videos.push(doc)
    },

    putPromise: (doc) => {
        return new Promise((resolve, reject) => {
            debug('putPromise', doc)
            self.put(doc, err => {
                if (err) {
                    reject(err)
                    return
                }

                debug('putPromise resolve')
                resolve()
            })

        })
    },

    // Set _id field of doc
    setId: (doc, tag, ext) => {
        let portion = self.getPortion()
        let _id = `${portion.name}/${self.name}`

        _id += '/' + doc.videoCreated
        if (doc.statusCreated)
            _id += ' ' + doc.statusCreated
        if (doc.noteCreated)
            _id += ' ' + doc.noteCreated
        if (doc.segmentCreated)
            _id += ' ' + doc.segmentCreated
        if (ext)
            _id += '.' + ext

        _id += `/${tag}`
        doc._id = _id
        doc.tag = tag
    },

    put: (doc, cb) => {
        debug('PUT', doc._id)

        let portion = self.getPortion()
        let db = portion.getDb()
        db.put(doc)
          .then(() => { cb && cb() })
          .catch(err => { cb && cb(err)})
    },

    setStatus: (passageVideo, status, cb) => {
        if (passageVideo.status === status) {
            cb && cb()
            return
        }

        let doc = {
            statusCreated: timestamp(),
            videoCreated: passageVideo.videoCreated,
            status,
        }
        self.statuses.push(doc)

        //debug('setStatus')
        //_.forEach(self.statuses, s => debug(s))
        
        self.setId(doc, 'status')
        self.put(doc, cb)
    },

    addVideo: (doc, cb) => {
        // doc: { username, videoCreated, duration, url }
        debug('addVideo', doc.duration, doc.url)
        //!! ensure that videoCreated is later than existing videos

        self.setId(doc, 'video', 'webm')
        self.videos.push(doc)

        self.put(doc, err => {
            if (err) {
                cb && cb(err)
                return
            }
            cb && cb(null, self.videos.slice(-1)[0])
        })
    },

    // Add a new note to this passage.
    // doc is actually the first segment to add to the new note.
    addNote: (doc, cb) => {
        let note = self.findOrCreateNote(doc)
        note.addSegment(doc, cb)
    },

    pushNote: note => {
        self.notes.push(note)
    },

    apply: (doc) => {
        //debug('passages apply', doc._id)
        
        let { noteCreated, _deleted } = doc

        if (noteCreated && _deleted) {
            self.applyDeleteNoteSegment(doc)
            return
        }
    
        let note
        if (noteCreated) note = self.findOrCreateNote(doc)
        
        switch (doc.tag) {
            case 'video': self.applyVideo(doc); break
            case 'status': self.applyStatus(doc); break
            case 'notesegment': note.apply(doc); break
            case 'resolved': note.apply(doc); break
            default: throw new Error(`Invalid tag: ${doc.tag}`)
        }
    },

    applyDeleteNoteSegment(doc) {
        let { noteCreated, segmentCreated } = doc
        let note = _.findWhere(self.notes, { noteCreated })
        if (!note) return

        let idx = _.findIndex(note.segments, { segmentCreated })
        if (idx < 0) return

        // we need to use detach rather than splice here because
        // splice immediately makes the object inaccessible and it is hard to
        // stop react from accessing it in that state and crashing
        detach(note.segments[idx])
    },

    applyVideo: (doc) => {
        let { videoCreated } = doc
        let video = _.findWhere(self.videos, { videoCreated })
        if (video) return   // Video already exists, nothing to do

        _insertBy(doc, self.videos, 'videoCreated')
    },

    applyStatus: (doc) => {
        let { statusCreated } = doc
        let status = _.findWhere(self.statuses, { statusCreated })
        if (status) return   // Status already exists, nothing to do

        _insertBy(doc, self.statuses, 'statusCreated')
    },

    findOrCreateNote: (doc) => {
        let { noteCreated } = doc
        let note = _.findWhere(self.notes, { noteCreated })
        if (note) return note

        let noteDoc = {
            videoCreated: doc.videoCreated,
            noteCreated: doc.noteCreated,
            position: doc.position,
            segments: [],
            resolved: false,
        }

        _insertBy(noteDoc, self.notes, 'position')
        
        note = _.findWhere(self.notes, { noteCreated })
        return note
    },

}))
.views(self => ({
    get status() {
        let lastPasageVideoStatus = _.last(self.statuses)
        return (lastPasageVideoStatus && lastPasageVideoStatus.status) || 0
    },

    get videosNotDeleted() {
        let videos = self.videos || []
        
        //debug('videos pre', videos.map(pv => `${pv.videoCreated}|${pv.status}`), deletedStatus)
        videos = _.filter(videos, pv => pv.status !== deletedStatus)
        //debug('videos post', videos.map(pv => `${pv.videoCreated}|${pv.status}`))
        return videos
    }
}))

