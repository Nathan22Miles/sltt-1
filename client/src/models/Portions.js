/*
     #item/<portion name>/portion
     #item/<portion name>/<passage name>/passage
 */ 

import { types, getParent } from "mobx-state-tree"
import _ from 'underscore'
import { detach } from "mobx-state-tree"

import { Passage } from './Passages.js'
import { _insertBy, removeFromDB } from './ModelUtils.js'

const debug = require('debug')('sltt:Portions') 


function _move(db, items, _id, newIndex, cb) { 
    let oldIndex = _.findIndex(items, { _id })
    if (oldIndex < 0) {
        cb(`_move: cannot find ${_id}`)
        return
    }

    if (oldIndex === newIndex) return

    // Moving an array element which is a model from array element oldIndex to newIndex is tricky. 
    // As far as I can tell you can't just use splice
    // to extract and then reinsert a model in mobx-state-tree array.
    // Removing the item from the array 'kills' it and then it cannot be resinerted.
    // Instead you need to get a plain js object representing the model and then reinsert that.
    let item = items[oldIndex].toJSON()
    items.splice(oldIndex, 1)
    items.splice(newIndex, 0, item)

    let prevRank = 0.0
    let nextRank = 200.0

    if (newIndex > 0) {
        prevRank = items[newIndex-1].rank
        nextRank = prevRank + 200.0
    }
    if (newIndex < items.length-1) { 
        nextRank = items[newIndex+1].rank
    }
    let newRank = (prevRank + nextRank) / 2.0
    items[newIndex].rank = newRank

    db.get(_id)
        .then(doc => {
            doc.rank = newRank
            return db.put(doc)
        })
        .then(() => cb && cb())
        .catch(err => { cb && cb(err) })
}

function _apply(doc, list) {
    if (doc._deleted) {
        let idx = _.findIndex(list, { _id: doc._id })
        if (idx < 0) return
        list.splice(idx, 1)
        return
    }

    _insertBy(doc, list, 'rank')
}

function _add(_db, doc, list, cb) {
    let rank = 100.0
    if (list.length > 0) rank = list[list.length - 1].rank + 100.0
    doc.rank = rank

    list.push(doc)

    _db.put(doc)
        .then(() => cb && cb())
        .catch(err => { 
            debug(err)
            cb && cb(err) 
        })
}

function _invalidNameCheck(name) {
    let match = name.match(/^[\w \-.]+$/)
    if (!match) return 'Must have only letters, numbers, hyphen(-), periods(.), and spaces'
    return null
}

function _duplicateCheck(items, name) {
    name = name.trim()
    if (_.findWhere(items, { name }))
        return 'Duplicate name'

    return null
}


// ================== PORTION =====================

// A portion is a related list of passages.


export const Portion = types.model("Portion", {
    _id: types.identifier(),
    name: types.string,
    // displayName: types.optional(types.string),
            // In order to allow renaming a portion without going back and changing all the references
            // to db items that mention the old name we will (someday) support a 'displayName' that
            // we show to the user.
    rank: types.number,  // order of this portion in the list of portions for this project
    passages: types.array(Passage, [])
})
.actions(self => ({
    load: () => {
        let options = {
            startkey: self.name + '/',
            endkey: self.name + '/\uffff',
            include_docs: true,
        }

        return new Promise((resolve, reject) => {
            self.getDb()
                .allDocs(options)
                .then(response => {
                    // response = {rows: [{doc: {...} }]}

                    let nonSegmentsRows = response.rows.filter(row => row.doc.tag !== 'segment')
                    let segmentRows = response.rows.filter(row => row.doc.tag === 'segment')

                    // process the segments last since they have to be applied individual
                    // VideoPassage entries
                    nonSegmentsRows.forEach(row => self.apply(row.doc))
                    segmentRows.forEach(row => self.apply(row.doc))

                    resolve()
                })
                .catch(err => {
                    let message = `${self.getName()} Portion.load ${err.stack}`
                    console.log(message)
                    debug('*** ' + message)
                    reject(message)
                })
        })
    },

    getDb() {
        let portions = getParent(self, 2)
        let _db = portions.getDb()
        return _db
    },

    // Accept a portions document from the database and apply it to the model
    apply: (doc) => {
        let parts = doc._id.split('/')
        
        if (parts[0] === '#item') {
            if (!doc.statuses) doc.statuses = [] // data migration
            _apply(doc, self.passages)
            return
        }

        let name = parts[1]
        let passage = _.findWhere(self.passages, { name })
        if (!passage) {
            debug(`${ self.getName() } *** no matching passage for ${doc._id}`)
            return
        }

        passage.apply(doc)
    },

    movePassage: (_id, newIndex, cb) => { 
        let _db = self.getDb()
        _move(_db, self.passages, _id, newIndex, cb) 
    },    

    removePassageReference(_id) {
        let project
        
        try {
            project = self.getProject()
        } catch (error) {
            // no parent, so no need to remove reference from parent
            // normally this should only happen when running unit tests
            return            
        }

        if (project.passage && project.passage._id === _id) {
            project.setPassage(null)
        }
    },

    removePassage: (_id, cb) => { 
        self.removePassageReference(_id)

        let idx = _.findIndex(self.passages, { _id })
        if (idx < 0) {
            cb(`_remove: cannot find ${_id}`)
            return
        }

        // we need to use detach rather than splice here because
        // splice immediately makes the object inaccessible and it is hard to
        // stop react from accessing it in that state and crashing
        detach(self.passages[idx])

        // Match this item all its subitems (if any).
        // E.g. for the portition #item/Progigal Son/portion
        // we also want to match the passage #item/Prodigal Son/Part 1/passage

        let itemsId = _id.slice(0, -('/passage'.length)) + '/'

        let videosId = itemsId.slice('#item/'.length)

        debug(`${self.getName()} removePassages ${itemsId}, ${videosId}`)

        let _db = self.getDb()

        removeFromDB(_db, videosId)
            .then(removeFromDB(_db, itemsId))
            .then(() => cb && cb())
            .catch(err => cb && cb(err))
    },    

    _addPassage: (name, cb) => {
        let err = self.checkName(name)
        if (err) {
            cb && cb(err)
            return
        }

        let _id = `#item/${self.name}/${name}/passage`
        let doc = { 
            _id, 
            name, 
            statuses: [],
            notes: [],
            videos: [],
        }
        _add(self.getDb(), doc, self.passages, cb)
    },
    
    addPassage: (name, cb) => {
        if (cb)
            return self._addPassage(name, cb)

        return new Promise((resolve, reject) => {
            self._addPassage(name, err => {
                if (err) reject(err)
                else resolve()
            })
        })
    },

    checkName: (name) => {
        return _invalidNameCheck(name) || _duplicateCheck(self.passages, name)
    },

    rename: (name, cb) => {
        cb('Renaming portions is not supported yet, sorry.')
        return

        // I think the following code is more or less right.
        // We would need to change all the ui components to use displayName instead of
        // name when then displayName has been set.

        // eslint-disable-next-line
        let err = self.checkName(name)
        if (err) {
            debug(`${self.getName()} *** rename ${err}`)
            cb && cb(err)
            return
        }

        let db = self.getDb()
        db.get(self._id)
            .then(doc => {
                doc.displayName = name
                return db.put(doc)
            })
            .then(() => cb && cb())
            .catch(err => { cb && cb(err) })
    },

    getProject: () => {
        let project = getParent(self, 3)
        return project
    },

    // Get name of portion for debugging purposes.
    // Don't crash no matter what.
    getName: () => {
        try {
            let project = self.getProject()
            return `[${project.name}/${self.name}]` 
        } catch (err) {
            return '*unknown*'
        }
    },

}))
.views(self => ({
    // true iff any passage in this portion has video already recorded
    get videod() {
        return self.passages.some(passage => passage.videod)
    },
}))


// ================== PORTIONS =======================

export const Portions = types.model("Portions", {
    portions: types.array(Portion, []),
})
.actions(self => {
    let _db

    return {
        load: () => {
            debug(`[${self.getProjectName()}] Portions#load`)

            let options = {
                startkey: '#item/',
                endkey: '#item/ufff0',
                include_docs: true,
            }

            let isPortion = row => (row.doc._id.endsWith('/portion'))
            let isPassage = row => (row.doc._id.endsWith('/passage'))

            return new Promise((resolve, reject) => {
                _db.allDocs(options)
                    .then(response => {
                        debug(`   [${self.getProjectName()}] Portions#load [rows=${response.rows.length}]`)
                        // response = {rows: [{doc: {...} }]}

                        // Process portions first so we have an owner for each passage
                        _.chain(response.rows)
                            .filter(isPortion)
                            .forEach(row => self.apply(row.doc))
                            
                        _.chain(response.rows)
                            .filter(isPassage)
                            .forEach(row => self.apply(row.doc))

                        resolve()
                    })
                    .catch(err => reject(err))
            }) 
        },

        apply: (doc) => {
            //debug('portions apply', doc._id)

            let parts = doc._id.split('/')
            let tag = parts[parts.length-1]

            if (parts[0] === '#item' && tag === 'portion') { 
                _apply(doc, self.portions)
                return
            }

            if (parts[0] === '#item' && tag === 'passage') {
                let portionName = doc._id.split('/')[1]
                
                // We inserted the portions first so we should 'always' find a matching 
                // portion for this passage
                let idx = _.findIndex(self.portions, { name: portionName })
                if (idx < 0) {
                    console.warn(`[${self.project.name}] Portions#apply item Portion not found [${doc._id}]`)
                    return
                }

                self.portions[idx].apply(doc)
                return
            }

            let name = parts[0]
            let portion = _.findWhere(self.portions, { name })
            if (!portion) {
                console.warn(`[${self.getProjectName()}] Portions#apply non item Portion not found [${doc._id}]`)
                return
            }

            portion.apply(doc)
        },

        setDb: (db) => { _db = db },

        getDb: () => { return _db },

        getProject: () => {
            let project = getParent(self, 1)
            return project
        },

        getProjectName: () => {
            try {
                return self.getProject().name
            } catch (err) {
                return '*unknown*'
            }
        },

        movePortion: (_id, newIndex, cb) => {
            _move(_db, self.portions, _id, newIndex, cb)
        },

        removePortionReference(_id) {
            let project

            try {
                project = self.getProject()
            } catch (error) {
                // no parent, so no need to remove reference from parent
                // normally this should only happen when running unit tests
                return
            }

            let portions = project.portions.portions

            if (project.portion && project.portion._id === _id) {
                if (portions.length > 0) {
                    project.setPortion(portions[0])
                } else {
                    project.setPortion(null)
                }
            }
        },

        removePortion: (_id, cb) => {
            self.removePortionReference(_id)

            let idx = _.findIndex(self.portions, { _id })
            if (idx < 0) {
                cb(`_remove: cannot find ${_id}`)
                return
            }

            // we need to use detach rather than splice here because
            // splice immediately makes the object inaccessible and it is hard to
            // stop react from accessing it in that state and crashing
            detach(self.portions[idx])

            // Match this item all its subitems (if any).
            // E.g. for the portition #item/Progigal Son/portion
            // we also want to match the passage #item/Prodigal Son/Part 1/passage

            let itemsId = _id.slice(0, -('/portion'.length)) + '/'

            let videosId = itemsId.slice('#item/'.length)

            debug(`[${self.getProjectName()}] Portions#removePortions`, itemsId, videosId)

            removeFromDB(_db, videosId)
                .then(removeFromDB(_db, itemsId))
                .then(() => cb && cb())
                .catch(err => cb && cb(err))
        },

        _addPortion: (name, cb) => {
            let err = self.checkName(name)
            if (err) {
                cb && cb(err)
                return
            }

            let _id = `#item/${name}/portion`
            let doc = { _id, name, passages: [] }
            _add(_db, doc, self.portions, cb)
        },

        addPortion: (name, cb) => {
            if (cb)
                return self._addPortion(name, cb)

            return new Promise((resolve, reject) => {
                self._addPortion(name, err => {
                    if (err) reject(err)
                    else resolve()
                })
            })
        },


        checkName: (name) => {
            return _invalidNameCheck(name) || _duplicateCheck(self.portions, name)
        }, 

    }
})
.views(self => ({
    get project() {
        return getParent(self, 1)
    },    
}))

