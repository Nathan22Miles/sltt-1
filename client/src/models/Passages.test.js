// Test Passages, Passage, PassageVideo, PassageNote, PassageNoteSegment models.
// Users in memory DB and should not access server.

import _ from 'underscore'
import PouchDB from 'pouchdb'

// Create an in-memory test DB.
// Exercise the model functionsâˆ« relating to passages.

import { Project } from './Project.js'
import { Portions, Portion } from './Portions.js'
import { Passages, Passage, PassageVideo, PassageNote, PassageNoteSegment } from './Passages.js'

require('../components/auth/User.js').user.setupTestUser()

// Create in-memory test DB
PouchDB.plugin(require('pouchdb-adapter-memory'))
let dbName = 'DB' + Math.random().toString(8)
let _test1 = new PouchDB(dbName, { adapter: 'memory' })

let portions, portion, passage

const project = Project.create({ name: 'ASLVtest' })


it("Can create empty Portions/portion/passage", (done) => {
    portions = Portions.create({ portions: [] })
    project.setPortions(portions)

    portions.setDb(_test1)

    portions.addPortion('Matthew', err => {
        expect(err).toBeFalsy()
        portion = portions.portions[0]

        portion.addPassage('Mat 1.1-5', err => {
            expect(err).toBeFalsy()
            passage = portion.passages[0] 

            portion.addPassage('Mat 1.6-8', err => {
                expect(err).toBeFalsy()
                done()
            })
        })
    })
})

it("Can add video", (done) => {
    let doc = {
        _id: 'Matthew/Mat 1.6-8/2018-01-01 00.00.00.webm/video',
        username: 'milesnlwork@gmail.com',
        videoCreated: '2018-01-01 00.00.00',
        url: 'Matthew/Mat 1.6-8/2018-01-01 00.00.00.webm',
        duration: 33.0,
    }

    passage.addVideo(doc, err => {
        if (err) throw new Error(err)
        expect(passage.videos.length).toBe(1)
        done()
    })
    
})

it("Can add video version 2", (done) => {
    let doc = {
        _id: 'Matthew/Mat 1.6-8/2018-01-02 00.00.00.webm/video',
        username: 'milesnlwork@gmail.com',
        videoCreated: '2018-01-02 00.00.00',
        url: 'Matthew/Mat 1.6-8/2018-01-02 00.00.00.webm',
        duration: 40.0,
    }

    passage.addVideo(doc, err => {
        if (err) throw new Error(err)
        expect(passage.videos.length).toBe(2)
        done()
    })

})

it("Can set status", (done) => {
    passage.setStatus(passage.videos[0], 2, err => {
        if (err) throw new Error(err)
        expect(passage.videos[0].status).toBe(2)

        passage.setStatus(passage.videos[1], 3, err => {
            if (err) throw new Error(err)
            expect(passage.videos[1].status).toBe(3)

            passage.setStatus(passage.videos[1], 4, err => {
                if (err) throw new Error(err)
                expect(passage.videos[1].status).toBe(4)

                done()
            })
        })
    })
})

it("Can add a new note", (done) => {
    let segment = {
        videoCreated:   '2018-01-01 00.00.00',
        noteCreated:    '2018-02-01 00.00.00',
        segmentCreated: '2018-04-01 00.00.00',
        username: 'milesnlwork@gmail.com',
        position: 1.0,
        duration: 22.0,
        url: 'https://blah2.com/blah.webm',
    }

    passage.addNote(segment, err => {
        if (err) throw new Error(err)
        expect(passage.notes.length).toBe(1)
        expect(passage.notes[0].segments.length).toBe(1)
        done()
    })
})

it("Can add a new note segment (out of date order)", (done) => {
    let segment = {
        videoCreated:   '2018-01-01 00.00.00',
        noteCreated:    '2018-02-01 00.00.00',
        segmentCreated: '2018-03-01 00.00.00',
        username: 'milesnlwork@gmail.com',
        position: 1.0,
        duration: 33.0,
        url: 'https://blah3.com/blah.webm',
    }

    passage.addNote(segment, err => {
        if (err) throw new Error(err)
        expect(passage.notes.length).toBe(1)
        expect(passage.notes[0].segments.length).toBe(2)
        expect(passage.notes[0].segments[0].segmentCreated).toBe('2018-03-01 00.00.00')
        done()
    })
})

it("Can resolve a note", (done) => {
    let note = passage.notes[0]

    note.resolve(err => {
        if (err) throw new Error(err)
        expect(passage.notes[0].resolved).toBe(true)
        done()
    })
})

it("Can remove a note segment", (done) => {
    let note = passage.notes[0]
    let _id = note.segments[1]._id

    note.removeSegment(_id, err => {
        if (err) throw new Error(err)
        expect(note.segments.length).toBe(1)
        done()
    })
})

// Prevent snapshot compare from failing due to running test at different time
function overrideStatusCreated(portion) {
    for (let passage of portion.passages) {
        for (let status of passage.statuses) {
            status.setStatusCreated('2018-03-15 00.00.00')
        }
    }
}

it("Can load passages from DB", (done) => {
    let portions2 = Portions.create({ portions: [] })

    portions2.setDb(_test1)

    portions2.load()
        .then(() => {
            let _portion = portions2.portions[0]
            return _portion.load()
        })
        .then(() => {
            overrideStatusCreated(portions2.portions[0])
            expect(portions2).toMatchSnapshot()
            done()
        })
        .catch(err => {
            expect(err).toBeFalsy()  // Fail quickly on error (to avoid wait for timeout)
        })
})
