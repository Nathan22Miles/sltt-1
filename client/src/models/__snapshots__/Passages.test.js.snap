// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Can load passages from DB 1`] = `
Object {
  "portions": Array [
    Object {
      "_id": "#item/Matthew/portion",
      "name": "Matthew",
      "passages": Array [
        Object {
          "_id": "#item/Matthew/Mat 1.1-5/passage",
          "name": "Mat 1.1-5",
          "notes": Array [
            Object {
              "noteCreated": "2018-02-01 00.00.00",
              "position": 1,
              "resolved": true,
              "segments": Array [
                Object {
                  "_id": "Matthew/Mat 1.1-5/2018-01-01 00.00.00 2018-02-01 00.00.00 2018-03-01 00.00.00.webm/notesegment",
                  "duration": 33,
                  "noteCreated": "2018-02-01 00.00.00",
                  "position": 1,
                  "segmentCreated": "2018-03-01 00.00.00",
                  "signedUrl": "",
                  "signedUrlExpires": 0,
                  "text": "",
                  "url": "https://blah3.com/blah.webm",
                  "username": "milesnlwork@gmail.com",
                  "videoCreated": "2018-01-01 00.00.00",
                },
              ],
              "videoCreated": "2018-01-01 00.00.00",
            },
          ],
          "rank": 100,
          "statuses": Array [
            Object {
              "status": 2,
              "statusCreated": "2018-03-15 00.00.00",
              "videoCreated": "2018-01-01 00.00.00",
            },
            Object {
              "status": 3,
              "statusCreated": "2018-03-15 00.00.00",
              "videoCreated": "2018-01-02 00.00.00",
            },
            Object {
              "status": 4,
              "statusCreated": "2018-03-15 00.00.00",
              "videoCreated": "2018-01-02 00.00.00",
            },
          ],
          "videos": Array [
            Object {
              "_id": "Matthew/Mat 1.1-5/2018-01-01 00.00.00.webm/video",
              "duration": 33,
              "signedUrl": "",
              "signedUrlExpires": 0,
              "url": "Matthew/Mat 1.6-8/2018-01-01 00.00.00.webm",
              "username": "milesnlwork@gmail.com",
              "videoCreated": "2018-01-01 00.00.00",
            },
            Object {
              "_id": "Matthew/Mat 1.1-5/2018-01-02 00.00.00.webm/video",
              "duration": 40,
              "signedUrl": "",
              "signedUrlExpires": 0,
              "url": "Matthew/Mat 1.6-8/2018-01-02 00.00.00.webm",
              "username": "milesnlwork@gmail.com",
              "videoCreated": "2018-01-02 00.00.00",
            },
          ],
        },
        Object {
          "_id": "#item/Matthew/Mat 1.6-8/passage",
          "name": "Mat 1.6-8",
          "notes": Array [],
          "rank": 200,
          "statuses": Array [],
          "videos": Array [],
        },
      ],
      "rank": 100,
    },
  ],
}
`;
